### PR TITLE
feat: get UTM codes from URL and add them to the app state and local …

### PR DESCRIPTION
**Pull request by anxolin**
# Summary

This PR gets UTM codes from URL and persist them into the local storage

This will add support for the following URL params
- utm_source
- utm_medium
- utm_campaign
- utm_content
- utm_term

Each parameter is independent and can be present or not. 
Also, it will clear the parameters from the URL once they are processed

This is connected to this PR https://github.com/cowprotocol/app-data/pull/25

Check the test section to see how to pass the params.

## Not included
This PR prepares for a follow up that will add this into the app-data based on https://github.com/cowprotocol/app-data/pull/25

# To Test

Use this URL:
https://swap-dev-git-utc-state-cowswap.vercel.app/#/swap?utm_source=twitter&utm_medium=email&utm_campaign=everyone-loves-cows-2023&utm_content=big-fat-button&utm_term=coincidence+of+wants

Check that the parameters will be updated in the local storage (only on page load, so if you want to override, you will need to open a new tab)

To do that, open devtools, and filter by UTM, you should see all the URL parameters there

image

They won't be lost on a refresh (only updated if new params are passed)

### CODE REVIEW

**Comment by bh2smith** [comment]
What is the expected size increase of unique app data records?

File: src/cow-react/modules/utm/hooks.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import { useSetAtom } from 'jotai'
+import { utmAtom } from './state'
+
+function getUtcParams(searchParams: URLSearchParams): UtmParams {
+  const utcSource = searchParams.get('utm_source') || undefined
+  const utcMedium = searchParams.get('utm_medium') || undefined
+  const utcCampaign = searchParams.get('utm_campaign') || undefined
+  const utcContent = searchParams.get('utm_content') || undefined
+  const utcTerm = searchParams.get('utm_term') || undefined
+
```
Comment (alfetopito) [inline src/cow-react/modules/utm/hooks.ts:13]: Shouldn't we also delete the params after loading them?
That's what's done for the other url params

**Comment by anxolin** [comment]
> What is the expected size increase of unique app data records?

@bh2smith What do you mean exactly? 

The app-data won't change, it will still be an IPFS hash. The content of the JSON will change. The formal definition is in this PR https://github.com/cowprotocol/app-data/pull/25

Basically all codes are not required (could be undefined), and they are just strings. There could be anything there

**Comment by bh2smith** [comment]
> What do you mean exactly?

I mean, will we now be creating a new (unique) app Data record for each order? In the past we have went to great lengths to ensure this did not grow so quickly and I am not sure if we are prepared to deal with this scale of growth.

**Comment by anxolin** [comment]
> I mean, will we now be creating a new (unique) app Data record for each order? 
No, we won't do that. But yes for anything that changes in this params. It's up to us to define the granularity we need. 
In this case, we can create a code per campaign or medium

> who will be responsible for publishing/pinning/uploading this content to IPFS in a timely manner? How long should we expect for this data to be considered "available"?

Same as before, the FE will make sure that is pinned and available. This might change though, since @nlordell might be moving this responsibility to the backend

File: src/cow-react/modules/utm/hooks.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      const searchParams = new URLSearchParams(search)
+      const utm = getUtmParams(searchParams)
+      console.log('UTM', utm, searchParams)
+      if (utm.utmCampaign || utm.utmCampaign || utm.utmContent || utm.utmMedium || utm.utmSource) {
+        // Only overrides the UTM if the URL includes any UTM param
+        setUtm(utm)
+      }
+
+      ALL_UTM_PARAMS.forEach((param) => searchParams.delete(param))
+      console.log('UTM 2', utm, searchParams)
```
Comment (alfetopito) [inline src/cow-react/modules/utm/hooks.ts:49]: Clear console logs?

**Review by alfetopito — APPROVED** [review]

File: src/cow-react/modules/utm/hooks.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+const UTM_TERM_PARAM = 'utm_term'
+
+const ALL_UTM_PARAMS = [UTM_SOURCE_PARAM, UTM_MEDIUM_PARAM, UTM_CAMPAIGN_PARAM, UTM_CONTENT_PARAM, UTM_TERM_PARAM]
+
+function getUtmParams(searchParams: URLSearchParams): UtmParams {
+  const utmSource = searchParams.get(UTM_SOURCE_PARAM) || undefined
+  const utmMedium = searchParams.get(UTM_MEDIUM_PARAM) || undefined
+  const utmCampaign = searchParams.get(UTM_CAMPAIGN_PARAM) || undefined
+  const utmContent = searchParams.get(UTM_CONTENT_PARAM) || undefined
+  const utmTerm = searchParams.get(UTM_TERM_PARAM) || undefined
```
Comment (alfetopito) [inline src/cow-react/modules/utm/hooks.ts:20]: What does `get` return if not found?

**Pull request merged**